romancal/skycell/skymap.py

```diff
@@ -0,0 +1,588 @@
+from scipy.spatial import KDTree
+from stcal.alignment import util as wcs_util
+
+from romancal.datamodels.library import ModelLibrary
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+
+SKYCELL_AREA = (4.6 / 3600.0) ** 2
+SKYTILE_AREA = (0.5) ** 2
+
+
+def image_coords_to_vec(
+    radec_coords: list[tuple[float, float]]
+    | tuple[list[float], list[float]]
+    | NDArray[float],
+) -> NDArray[float]:
+    """
+    This routine can handle the corners in both organizations, whether
+    a sequence of ra, dec pairs or a list of ra coordinates, followed
+    by a list of dec coordinates. So long as either form can be converted
+    to a numpy array it is ok. The spherical geometry routines expect
+    the latter organization, and the returned value will be of that
+    organization.
+    """
+
+    radec_coords = np.array(radec_coords)
+    if radec_coords.shape == (2, 4):
+        radec_coords = radec_coords.transpose()
+
+    # Convert all celestial coordinates to cartesion coordinates.
+    return np.stack(
+        sgv.lonlat_to_vector(radec_coords[:, 0], radec_coords[:, 1]), axis=1
+    )
+
+
+class SkyCell:
+    __index: int | None
+    __data: np.void
+
+    def __init__(self, index: int | None):
+        self.__index = index
+        if index is not None:
+            self.__data = SKYMAP.skycells[index]
+
+    @classmethod
+    def from_name(cls, name: str):
+        if not re.match(r"\d{3}\w\d{2}x\d{2}y\d{2}", name):
+            raise ValueError(f"invalid skycell name {name}")
+        return SkyCell(np.where(SKYMAP.skycells["name"] == name)[0][0])
+
+    @classmethod
+    def from_data(cls, data: np.void) -> "SkyCell":
+        instance = cls(index=None)
+        instance.__data = data
+        return instance
+
+    @classmethod
+    def from_center_and_coordinates(
+        cls,
+        skytile_center: tuple[float, float],
+        skycell_coordinates: tuple[float, float],
+    ) -> "SkyCell":
+        return cls.from_name(
+            f"r{round(skytile_center[0]):03}d{'p' if skytile_center[1] >= 0 else 'm'}{round(skytile_center[1]):02}x{'p' if skycell_coordinates[1] >= 0 else 'm'}{skycell_coordinates[1]:02}y{'p' if skycell_coordinates[1] >= 0 else 'm'}{skycell_coordinates[1]:02}"
+        )
+
+    @classmethod
+    def from_modellibrary(cls, library: ModelLibrary) -> "SkyCell":
+        if (
+            "skycell_wcs_info" in library.asn
+            and library.asn["skycell_wcs_info"] != "none"
+        ):
+            skycell_name = library.asn["skycell_wcs_info"]["name"]
+        elif "target" in library.asn:
+            # check to see if the product name contains a skycell name & if true get the skycell record
+            skycell_name = library.asn["target"]
+        else:
+            raise ValueError(
+                "cannot extract skycell information from modellibrary association with neither WCS nor target info"
+            )
+
+        return SkyCell(skycell_name)
```

Comment:
Should this be `SkyCell.from_name(skycell_name)`? At least that's what works for me when passing a sky cell name. Also it may be good to add a test.